libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -192,11 +223,118 @@ bool AP_OABendyRuler::update(const Location& current_loc, const Location& destin
+    const uint8_t limit = 180 / OA_BENDYRULER_BEARING_INC_VERTICAL;
+
+    for (uint8_t i = 0; i <= limit; i++) {
+        for (uint8_t bdir = 0; bdir <= 1; bdir++) {
+            // skip duplicate check of bearing straight towards destination or 180 degrees behind
+            if (((i==0) && (bdir > 0)) || ((i == limit) && (bdir > 0))) {
+                continue;
+            }
+
+            // bearing that we are probing
+            const float vertical_bearing_delta = i * OA_BENDYRULER_BEARING_INC_VERTICAL * (bdir == 0 ? 1.0f : -1.0f);
+
+            Location test_loc = current_loc;
+            test_loc.offset_bearing_and_pitch(bearing_to_dest, vertical_bearing_delta,lookahead_step1_dist);
+
+            // calculate margin from fence for this scenario
+            float margin = calc_avoidance_margin(current_loc, test_loc);
+
+            if (margin > best_margin) {
+                best_margin_pitch = vertical_bearing_delta;
+                best_margin = margin;
+            }
+
+            if (margin > _margin_max) {
+                //This path avoids the obstacles with the required margin, now check for the path ahead
+                if (!have_best_vertical_bearing) {
+                    best_vertical_bearing = vertical_bearing_delta;
+                    have_best_vertical_bearing = true;
+                }
+                const float test_vertical_bearings_step2[] { 0.0f, 90.0f, -90.0f, 180.0f};
+                float bearing_to_dest2;
+                is_equal(fabsf(vertical_bearing_delta), 90.0f) ? (bearing_to_dest2 = bearing_to_dest) : (bearing_to_dest2 = test_loc.get_bearing_to(destination) * 0.01f);
+                float distance2 = constrain_float(lookahead_step2_dist, OA_BENDYRULER_LOOKAHEAD_STEP2_MIN, test_loc.get_distance(destination));
+
+                for (uint8_t j = 0; j < ARRAY_SIZE(test_vertical_bearings_step2); j++) {
+                    float bearing_test2 = wrap_180(test_vertical_bearings_step2[j]);
+                    Location test_loc2 = test_loc;
+                    test_loc2.offset_bearing_and_pitch(bearing_to_dest2,bearing_test2 ,distance2);
+
+                    // calculate minimum margin to fence and obstacles for this scenario
+                    float margin2 = calc_avoidance_margin(test_loc, test_loc2);
+                    if (margin2 > _margin_max) {
+                        // all good, now project in the chosen direction by the full distance
+                        destination_new = current_loc;
+                        destination_new.offset_bearing_and_pitch(bearing_to_dest,vertical_bearing_delta, distance_to_dest);
+                        _current_lookahead = MIN(_lookahead, _current_lookahead * 1.1f);
+                        // if the chosen direction is directly towards the destination turn off avoidance
+                        // i == 0 && j == 0 implies no deviation from bearing to destination
+                        const bool active = (i != 0 || j != 0);
```

Comment:
Ok, finally a real question.. does this correctly turn off avoidance in the situation below?  I.e. will it turn off avoidance when the vehicle reaches 2?  or will it prematurely turn off when it reaches 1?  I'm just slightly concerned that the 2nd probe's bearing is directly towards the final destination horizontally but perhaps not vertically?
![image](https://user-images.githubusercontent.com/1498098/89413020-068f2500-d763-11ea-87a5-6fa3e843d037.png)

Reply:
@rmackay9 thanks for pointing this out to me, I missed it! 
Yes you're correct, so what my PR will do is, climb till it clears of the margin to the obstacle.. So it will turn off avoidance at 1. In simulations and real life testing this seemed to work out great. I have been working towards a fix where, if we were climbing up to avoid an obstacle , in step 2 we will first look downwards (-90)... And if we were descending while BendyRuler was active, in step 2 we will look up (+90) first. 
However, this does require some additional logic and messing around. I have completed the rest the changes that you have asked, do you like this new logic, or do you have something else in mind?

Reply:
I think that logic makes sense.  My first thought was that we should look directly towards the destination (i.e. if the vehicle is at 1 then it looks down and forward towards the destination) but we never fly vertically diagonally because we're unlikely to have a sonar pointing in that direction so I think looking down is the only option.